reqresp/workflow.go

```diff
@@ -0,0 +1,195 @@
+	// be too large before continue-as-new. This map is only here to serve queries
+	// and could be removed in a response-activity-only setup.
+	responses               map[string]*Response
+	responseActivityOptions workflow.ActivityOptions
+}
+
+func newUppercaser(ctx workflow.Context) (*uppercaser, error) {
+	u := &uppercaser{
+		// For the main context, we're only going to allow 1 retry and only a 5
+		// second schedule-to-close timeout. WARNING: The timeout and retry affect
+		// how long this workflow stays open and may prevent it from performing its
+		// continue-as-new until timeout occurs and/or retries are finished.
+		Context: workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			ScheduleToCloseTimeout: 5 * time.Second,
+			RetryPolicy:            &temporal.RetryPolicy{MaximumAttempts: 2},
+		}),
+
+		// Use a signal to obtain requests. Since this cannot be closed, the
+		// workflow counts on the ability to have some "idle" period between
+		// handling requests/responses where it can continue-as-new.
+		requestCh: workflow.GetSignalChannel(ctx, "request"),
+
+		// We'll allow 500 requests before we continue-as-new the workflow. This is
+		// required because the history will grow very large otherwise for an
+		// interminable workflow fielding signal requests and executing activities.
+		requestsBeforeContinueAsNew: 500,
+		responses:                   make(map[string]*Response, 500),
+
+		// We're gonna just support 10 second timeout and 3 retries on response
+		// callback activities. WARNING: The timeout and retry affect how long this
+		// workflow stays open and may prevent it from performing its
+		// continue-as-new until timeout occurs and/or retries are finished.
+		responseActivityOptions: workflow.ActivityOptions{
+			// We use schedule-to-close because if the requester side is not present,
+			// this may hang otherwise with just start-to-close.
+			ScheduleToCloseTimeout: 10 * time.Second,
+			RetryPolicy:            &temporal.RetryPolicy{MaximumAttempts: 4},
+		},
+	}
+	// Set query handler
+	if err := workflow.SetQueryHandler(ctx, "response", u.queryResponse); err != nil {
+		return nil, fmt.Errorf("failed setting query handler: %w", err)
+	}
+	return u, nil
+}
+
+func (u *uppercaser) run() error {
+	// We create one selector for use for the entire run. Since we do a
+	// continue-as-new after the workflow gets too large, continually adding
+	// futures is ok.
+	selector := workflow.NewSelector(u)
+
+	// Listen for cancelled
+	var cancelled bool
+	selector.AddReceive(u.Done(), func(workflow.ReceiveChannel, bool) { cancelled = true })
+
+	// Listen for new requests
+	var requestCount int
+	selector.AddReceive(u.requestCh, func(c workflow.ReceiveChannel, more bool) {
+		requestCount++
+		var req Request
+		c.Receive(u, &req)
+		u.addExecuteActivityFuture(&req, selector)
+	})
+
+	// Continually select until there are too many requests and no pending selects
```

Comment:
I would expand the "no pending selects" comment. The "draining" is mentioned in some other comment, but this is exactly what guarantees it.

Reply:
Will do. In fact, I've found the single-central-selector pattern to be so valuable, I am wondering if we should document this "reactor pattern" and encourage its use more (reminds me of actors).

Reply:
Expanded.